redesign learner page

aliceriot opened the pull request:
#### What are the relevant tickets?

closes #3355 

#### What's this PR do?

this is a pretty simple change the adapts the `CourseListCard` component for the Learner page by adding a `showStaffView` prop that hides certain pieces of information, changes the title, etc.

#### How should this be manually tested?

take a look at the learner page, and make sure you see some details about programs you are enrolled in. you will need to be a staff user for the card to show up.

noisecapella commented:
I'm getting an error:

![screenshot from 2017-07-06 15-10-25](https://user-images.githubusercontent.com/863262/27928452-4a3c7f24-625d-11e7-8a88-0dedd89e965b.png)

I think it happens when `currentProgramEnrollment` is not set, which is true briefly when a user is first accessing the site. Can you add a check in `FinancialAidCalculator` in case it's null?

noisecapella commented:
Some of the states don't have messages like they do in the dashboard. For example this is for a failed course:

![screenshot from 2017-07-06 16-25-54](https://user-images.githubusercontent.com/863262/27931334-dc070802-6267-11e7-8c82-95df61e1d5ef.png)

noisecapella commented:
Nevermind, looks like this is intentional per the issue

noisecapella commented:
Here's the screenshots of the various dashboard states:

[dashboard_states_pr_3373.zip](https://github.com/mitodl/micromasters/files/1129217/dashboard_states_pr_3373.zip)

roberthouse54 commented:
This needs to show a status text under the grades, made up of two parts, separated by a dash. The first part shows whether the course is currently in progress, or not yet started, or ended. The second part shows whether the user is Auditing (with payment deadline date), Paid, Audited (if the course has ended), or Paid but not enrolled (which is an error state). See mock below.

![image](https://user-images.githubusercontent.com/20047260/27971588-3dabae8c-6321-11e7-8487-75c0ad7a0dfd.png)

noisecapella commented on static/js/components/Learner.js:
```diff
@@ -62,12 +63,21 @@ export default class Learner extends React.Component {
-          setShowGradeDetailDialog={setShowGradeDetailDialog}
-          dialogVisibility={ui.dialogVisibility}
-        />
+        <div key={program.title}>
```
Can we use `program.id` here instead? `title` isn't technically guaranteed to be unique

aliceriot commented:
@noisecapella I believe I addressed all of what needs to happen (the new small course message display). It should now have what's described in the issue.

noisecapella commented:
Here's the screenshot we were looking at:

dashboard:
![dashboard_state_004_set_to_passed](https://user-images.githubusercontent.com/863262/28230675-f916237a-68b5-11e7-81b8-6aa6a12916df.png)

learner:
![dashboard_state_004_set_to_passed](https://user-images.githubusercontent.com/863262/28230681-0095512a-68b6-11e7-97ba-55e98893069e.png)

noisecapella commented:
For `014_set_past_run_to_passed` there is an auditing upgrade message in the dashboard

![dashboard_state_014_set_past_run_to_passed](https://user-images.githubusercontent.com/863262/28232400-7a377e60-68be-11e7-900c-fc3232631474.png)


but there's no auditing message in the learner page:

![dashboard_state_014_set_past_run_to_passed](https://user-images.githubusercontent.com/863262/28232341-2719d3e0-68be-11e7-952d-73047df8a82c.png)

noisecapella commented:
If the user passes the edX course but doesn't pass the exam, they see this message on the dashboard:

![dashboard_state_018_create_exams_edx_ _exam_ _with_new_offering](https://user-images.githubusercontent.com/863262/28232500-10fa3630-68bf-11e7-9259-24bec5a567bd.png)

but they only see 'Passed' on the learner page:

![dashboard_state_018_create_exams_edx_ _exam_ _with_new_offering](https://user-images.githubusercontent.com/863262/28232513-1cb7990e-68bf-11e7-82c9-415a2bbc394a.png)

aliceriot commented:
for the first case (the auditing message) do you have the JSON for the API response? I'm not exactly sure how to address that case.

aliceriot commented:
@noisecapella bump

pdpinch commented:
What's the status of this?

aliceriot commented:
still in review

noisecapella commented:
Sorry for the delay, I'll go through the screenshot differences and paste the ones with the JSON files in separate comments

noisecapella commented:
`004_set_to_passed` has a dash on the dashboard page without any text. On master it isn't there:

this branch:
![dashboard_state_004_set_to_passed](https://user-images.githubusercontent.com/863262/28683955-981cae3e-72d0-11e7-9aa1-b32177a61f69.png)

[dashboard_state_004_set_to_passed.zip](https://github.com/mitodl/micromasters/files/1180762/dashboard_state_004_set_to_passed.zip)

master:
[dashboard_state_004_set_to_passed-master.zip](https://github.com/mitodl/micromasters/files/1180766/dashboard_state_004_set_to_passed-master.zip)

noisecapella commented:
There are similar changes for other dashboard screenshots. Running through the learner page screenshots now...

noisecapella commented:
I'm still seeing this error on occasion when running through the screenshots: https://github.com/mitodl/micromasters/pull/3373#issuecomment-313491867

noisecapella commented:
Here's all of the screenshots I took with the API JSON data, for both the dashboard page and the learner page. I'll comment again with the differences I noticed

[screenshots.zip](https://github.com/mitodl/micromasters/files/1180847/screenshots.zip)

noisecapella commented:
- In the dashboard page for 001_set_to_enrolled there's a dash after `Course in progress -` which isn't there in master.
 - In the learner page for 004_set_to_passed there is a `- Passed` without anything before the dash. In the dashboard page there is a dash where there wasn't anything before in master
 - In the learner page for 014_set_past_run_to_passed it used to say `Auditing` in master but in this PR it doesn't say anything
 - In the learner page for 016 it says `- Passed` while also having the icon for `Passed` (not sure if this matters). On master it says `Paid` instead
 - In the learner page for 024_failed_with_prev_passed_run you have a passed icon but the text says `- Paid, did not pass`
 - In 037_missed_payment_can_reenroll the dashboard shows a 'Pay now' button but the learner page says `Course in progress - Missed payment deadline`

aliceriot commented:
@noisecapella I can't seem to reproduce that error (with the financial aid calculator), do you have any ideas about what is causing it?

aliceriot commented:
ok, going through and addressing the screenshots:

- 001_set_to_enrolled: the draft isn't showing up locally for me:

![no-dash](https://user-images.githubusercontent.com/6207644/29079246-a3cbf5b0-7c2a-11e7-8234-94439ce66ce8.png)

 - 004_set_to_passed: there was a `- Passed` without anything before the dash, just fixed it:

![passssseed](https://user-images.githubusercontent.com/6207644/29079597-9216a74c-7c2b-11e7-9efe-498b4761d1fa.png)

 - In the learner page for 014_set_past_run_to_passed it used to say `Auditing` in master but in this PR it doesn't say anything

I looked at the JSON, the issue for this one is that the `end_date` for the course is in the past. we only want to show the 'Auditing' message if the user is currently auditing an ongoing course, I think.

     
 - In the learner page for 016 it says `- Passed` while also having the icon for `Passed` (not sure if this matters). On master it says `Paid` instead

I think this is the intended behavior, the designs don't show any changes to the grades component (what draws the icon) for the learner page.

 - In the learner page for 024_failed_with_prev_passed_run you have a passed icon but the text says `- Paid, did not pass`

Just fixed this, just a little logical error. The function that returns 'Paid, did not pass' was only checking if the most recent course run was passed, where it should be checking all course runs.

 - In 037_missed_payment_can_reenroll the dashboard shows a 'Pay now' button but the learner page says `Course in progress - Missed payment deadline`

it looks like that's the existing behavior on the dashboard, but that doesn't seem right (why would we should 'pay now' if you missed the deadline?). Do you think we should address that here? (or am I misunderstanding something)

noisecapella commented:
@aliceriot Regarding the financial aid calculator exception it happened during the `snapshot_dashboard_states` command. I'll run it again and let you know if it's still a problem

noisecapella commented:
I filed a separate PR #3458 for the null reference thing

noisecapella commented:
On the dashboard page in this branch I'm seeing `Course in progress -`:
![dashboard_state_000_set_to_enrolled](https://user-images.githubusercontent.com/863262/29193549-d9285b7a-7df3-11e7-8fe4-332d990885b4.png)

noisecapella commented:
Other than that everything looks great. I filed https://github.com/mitodl/micromasters/issues/3406 for the dashboard error, probably out of scope for here

aliceriot commented:
ok, I think I fixed it!

aliceriot commented:
alright, here we go
